cpp/src/arrow/device.h

```diff
@@ -105,6 +107,63 @@ class ARROW_EXPORT Device : public std::enable_shared_from_this<Device>,
   bool is_cpu_;
 };
 
+/// \brief EXPERIMENTAL: An object that provides event/stream sync primitives
+///
+///
+class ARROW_EXPORT DeviceSync {
+ public:
+  explicit DeviceSync(void* sync_event) : sync_event_{sync_event}, owns_event_{false} {}
```

Comment:
This constructor should probably be protected; IIUC we'll always be constructing a device-specific derived class and never a bare DeviceSync

Reply:
that's a good idea